[WIP] header: New HeaderMatcher type - Contains

shivanshu21 opened the pull request:
Fixes: 12590

Signed-off-by: Shivanshu Goswami <shigoswami@ebay.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message: header: New HeaderMatcher type - Contains
Additional Description: For matching values in the header that might be somewhere in the middle of the header, the present option is to use Regex in the form .*Search-Pattern.*. This can cause catastrophic backtracking as described in #7728
As a solution, I have introduced another header match type called `contains` which is based on absl::StrContains().

Risk Level: Low
Testing: Unit tests are included and manual testing was performed.
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

shivanshu21 commented:
Tests passing in local
================

```
shigoswami@LM-SJL-11016036 ~/gitcode/envoy: /private/vbazel test -c opt //test/common/http:header_utility_test
INFO: Analyzed target //test/common/http:header_utility_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
Target //test/common/http:header_utility_test up-to-date:
  bazel-bin/test/common/http/header_utility_test
INFO: Elapsed time: 14.771s, Critical Path: 13.52s
INFO: 3 processes: 3 darwin-sandbox.
INFO: Build completed successfully, 4 total actions
//test/common/http:header_utility_test                                   PASSED in 0.1s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 4 total actions
shigoswami@LM-SJL-11016036 ~/gitcode/envoy: 
```

lizan commented on api/envoy/api/v2/route/route_components.proto:
```diff
@@ -1568,6 +1568,15 @@ message HeaderMatcher {
     //
     // * The suffix *abcd* matches the value *xyzabcd*, but not for *xyzbcd*.
     string suffix_match = 10 [(validate.rules).string = {min_bytes: 1}];
+
+    // If specified, header match will be performed based on whether the header value contains
+    // the given value or not.
+    // Note: empty contains match is not allowed, please use present_match instead.
+    //
+    // Examples:
+    //
+    // * The value *abcd* matches the value *xyzabcdpqr*, but not for *xyzbcdpqr*.
+    string contains_match = 12 [(validate.rules).string = {min_bytes: 1}];
```
v2 is deprecated and freezed, no change is allowed in v2.

shivanshu21 replied:
Removed v2 api changes.

htuch commented on api/envoy/api/v2/route/route_components.proto:
```diff
@@ -1568,6 +1568,15 @@ message HeaderMatcher {
     // * The suffix *abcd* matches the value *xyzabcd*, but not for *xyzbcd*.
     string suffix_match = 10 [(validate.rules).string = {min_bytes: 1}];
+
+    // If specified, header match will be performed based on whether the header value contains
```
I wonder if we just want to add support general globs, rather than special casing this.

shivanshu21 replied:
@htuch This is being added for issue opened here: https://github.com/envoyproxy/envoy/issues/12590

This is no more a special case than `suffix-match` or `present-match`. @htuch and @mattklein123 Later today I will add the same matcher for StringMatcher as well.  Please let me know if we want to proceed this way.

Is there another PR or design doc available for wildcard based matches? Regex matching with `std::regex` can cause catastrophic backtracking and occasional crashes. Is there such a possibility with glob/wildcard matching too? Substring matches are O(n) as they are based on KMP.

shivanshu21 replied:
@htuch I did a little more digging. I see that the `Xray` tracer extension has a wildcard match:
https://github.com/envoyproxy/envoy/pull/8526/files#diff-620892d553c2be71cf26761d0c9db408R26

I am still breaking it down but right now it looks like it supports `?` and `*` which is good enough for our case. 

Do you want just these two wildcards or a full set? If you're okay with this I can refactor the code a little to reuse the same matcher.

htuch replied:
I think we would want a syntax more like https://github.com/envoyproxy/envoy/issues/7763. Maybe we should go with a simple contains for now.

htuch commented:
@shivanshu21 one other question, do you get catastophic backtracing with RE2 for this case? Do you know what the performance penalty of using RE2 vs. a strcontains operation is here?

shivanshu21 commented:
> @shivanshu21 one other question, do you get catastophic backtracing with RE2 for this case? Do you know what the performance penalty of using RE2 vs. a strcontains operation is here?

@htuch We haven't tried with Safe Regex as the definition says this:

```
(type.matcher.v3.RegexMatcher) If specified, this regex string is a regular expression rule which implies the entire request header value must match the regex. The rule will not match if only a subsequence of the request header value matches the regex.
```

The rule will not match if only a subsequence of the header matches the regex. So looks like strContains() case won't be handled here.

We have however tried comparing `deprecated` regex to RE2 and RE2 outperforms significantly.

shivanshu21 commented:
Matchers test passing in local
------------------------------

```
shigoswami@LM-SJL-11016036 ~/gitcode/envoy: bazel test -c opt //test/common/common:matchers_test
INFO: Analyzed target //test/common/common:matchers_test (0 packages loaded, 0 targets configured).
INFO: Found 1 test target...
Target //test/common/common:matchers_test up-to-date:
  bazel-bin/test/common/common/matchers_test
INFO: Elapsed time: 12.590s, Critical Path: 11.52s
INFO: 3 processes: 3 darwin-sandbox.
INFO: Build completed successfully, 4 total actions
//test/common/common:matchers_test                                       PASSED in 0.1s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 4 total actions
```

htuch commented:
RE2 should work, as long as you glob at the start and end, e.g. ".*<string>.*".

htuch commented:
LGTM, but just want to check with other @envoyproxy/api-shepherds that this makes sense. I think that RE2 is not the right answer for performance here, but do we want to essentially open up arbitrary string predicates as first class API elements?

shivanshu21 commented:
Thanks! I will wait for API shepherds to respond before merging.

shivanshu21 commented:
@lizan PTAL

shivanshu21 commented:
@envoyproxy/api-shepherds , @lizan This PR has been awaiting review. Please take a look.

lizan commented on source/common/common/matchers.cc:
```diff
@@ -100,6 +100,10 @@ bool StringMatcherImpl::match(const absl::string_view value) const {
                                   : absl::EndsWith(value, matcher_.suffix());
+  case envoy::type::matcher::v3::StringMatcher::MatchPatternCase::kContains:
+    return matcher_.ignore_case() ? absl::StrContains(absl::AsciiStrToLower(value),
+                                                      absl::AsciiStrToLower(matcher_.contains()))
```
can we cache the lowered matcher_.contains somewhere for this case?

shivanshu21 commented:
Matchers and Header utility tests passing
------------------------------------------

```
Target //test/common/common:matchers_test up-to-date:
  bazel-bin/test/common/common/matchers_test
INFO: Elapsed time: 346.251s, Critical Path: 70.60s
INFO: 359 processes: 359 darwin-sandbox.
INFO: Build completed successfully, 385 total actions
//test/common/common:matchers_test                                       PASSED in 0.2s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 385 total actions
shigoswami@LM-SJL-11016036 ~/gitcode/envoy:
```

```
shigoswami@LM-SJL-11016036 ~/gitcode/envoy: bazel test -c opt //test/common/http:header_utility_test
INFO: Analyzed target //test/common/http:header_utility_test (1 packages loaded, 2 targets configured).
INFO: Found 1 test target...
Target //test/common/http:header_utility_test up-to-date:
  bazel-bin/test/common/http/header_utility_test
INFO: Elapsed time: 17.793s, Critical Path: 17.06s
INFO: 3 processes: 3 darwin-sandbox.
INFO: Build completed successfully, 4 total actions
//test/common/http:header_utility_test                                   PASSED in 0.1s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 4 total actions
```

@lizan I have added the lowercase string to the class so that it is only computed once. PTAL.

shivanshu21 commented:
/retest

dio commented on docs/root/version_history/current.rst:
```diff
@@ -13,6 +13,7 @@ Minor Behavior Changes
 
 * compressor: always insert `Vary` headers for compressible resources even if it's decided not to compress a response due to incompatible `Accept-Encoding` value. The `Vary` header needs to be inserted to let a caching proxy in front of Envoy know that the requested resource still can be served with compression applied.
 * decompressor: headers-only requests were incorrectly not advertising accept-encoding when configured to do so. This is now fixed.
+* http: added a new header matcher type called ``contains`` which matches if the value of the header has the substring mentioned in contains matcher.
```
Usually, we link the added field here. In this case, probably:

```rst
* matcher: added :ref:`contains <envoy_v3_api_field_type.matcher.v3.StringMatcher.contains>` to ...
```

dio replied:
And, seems you have another one for `route` as well.

shivanshu21 commented:
@mattklein123 How do we remove the `v2 api` blocker labels? Merging is blocked because of that.

@lizan and Htuch have both LGTMed and Lizan has LGTMed for API changes as well. 
There are `No v2 api` changes in this PR.

I have addressed all the comments including adding release notes with proper links.

mattklein123 commented on api/envoy/type/matcher/string.proto:
```diff
@@ -16,7 +16,7 @@ option (udpa.annotations.file_status).package_version_status = FROZEN;
 
 // Specifies the way to match a string.
-// [#next-free-field: 7]
+// [#next-free-field: 8]
```
This is still modifying a v2 API. Please revert this.

shivanshu21 replied:
Oh! I thought all the v2 apis are in `api/envoy/api/v2` folder. Reverting.

shivanshu21 commented:
@mattklein123 

Addressed the comment. Please take a look and suggest ways to unblock the merge.
